File: sparktrain/src/main/scala/com/xtb/spark/dao/CourseClickCountDao.scala
package com.xtb.spark.dao

import com.xtb.spark.domain.CourseClickCount
import com.xtb.spark.utils.HBaseUtils
import org.apache.hadoop.hbase.client.Get
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.mutable.ListBuffer

/**
  * 实战课程的点击数数据访问层
  */
object CourseClickCountDao {
  val TABLE_NAME = "imooc_course_clickcount"
  val CF = "info"
  val QUALIFIER = "click_count"


  /**
    * 保存数据到HBase
    * @param list CourseClickCount集合
    */
  def save(list: ListBuffer[CourseClickCount]): Unit = {
    val table = HBaseUtils.getInstance().getTable(TABLE_NAME)
    for (ele <- list) {
      // incrementColumnValue这个API是增量列值的意思, 将value数值进行累加
      table.incrementColumnValue(Bytes.toBytes(ele.day_course), Bytes.toBytes(CF), Bytes.toBytes(QUALIFIER), ele.click_count)
    }
  }

  /**
    * 根据rowKey查询数据
    * @param day_course 就是rowKey的值 --> 20180502_130
    * @return 返回对应的20180502_130的访问总数
    */
  def count(day_course: String): Long = {
    val table = HBaseUtils.getInstance().getTable(TABLE_NAME)
    val get = new Get(Bytes.toBytes(day_course))
    val value = table.get(get).getValue(CF.getBytes(), QUALIFIER.getBytes)
    if (null == value) {
      0L
    } else {
      Bytes.toLong(value)
    }
  }

  def main(args: Array[String]): Unit = {
    val list = new ListBuffer[CourseClickCount]
    list.append(CourseClickCount("20171111_8", 8))
    list.append(CourseClickCount("20171111_9", 9))
    list.append(CourseClickCount("20171111_1", 100))
    save(list)
    println(count("20171111_8") + " : " + count("20171111_9") + " : " + count("20171111_1"))
  }
}

File: spark-web/src/main/java/com/xtb/spark/sparkweb/dao/CourseClickCountDao.java
package com.xtb.spark.sparkweb.dao;

import com.xtb.spark.sparkweb.domain.CourseClickCount;
import com.xtb.spark.sparkweb.utils.HBaseUtils;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * 实战课程访问量数量数据访问层
 */
public class CourseClickCountDao {

    /**
     * 根据天查询HBase数据
     */
    public List<CourseClickCount> query(String tableName, String day) throws IOException {
        // 去HBase表中根据day去获取实战课程的点击数
        Map<String, Long> map = HBaseUtils.getInstance().query(tableName, day);
        return map.entrySet().stream()
                .map(x -> CourseClickCount.of().setName(x.getKey()).setValue(x.getValue()))
                .collect(Collectors.toList());
    }
}
